File: StreamProcessorService/src/main/java/com/stream/processor/enrichment/CompositeEnrichmentStrategy.java
package com.stream.processor.enrichment;

import com.stream.processor.Model.SensorEvent;
import lombok.RequiredArgsConstructor;

import java.util.List;

@RequiredArgsConstructor
public class CompositeEnrichmentStrategy implements EnrichmentStrategy {

    private final List<EnrichmentStrategy> strategies ;

    @Override
    public SensorEvent enrich(SensorEvent event) {
        SensorEvent enriched = event ;
        for(EnrichmentStrategy strategy : strategies) {
            enriched = strategy.enrich(enriched);
        }
        return enriched;
    }
}

File: validation_service/composite_enrichment_strategy.py
"""Applies multiple enrichment strategies"""
import sys
import os
from typing import List, Dict, Any

# Add current directory to path for imports
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from hrv_enrichment_strategy import HrvEnrichmentStrategy
from bmi_enrichment_strategy import BmiEnrichmentStrategy
from activity_level_enrichment_strategy import ActivityLevelEnrichmentStrategy
from timestamp_enrichment_strategy import TimestampEnrichmentStrategy


class CompositeEnrichmentStrategy:
    """Applies multiple enrichment strategies"""
    
    def __init__(self, strategies: List = None):
        if strategies is None:
            self.strategies = [
                HrvEnrichmentStrategy(),
                BmiEnrichmentStrategy(),
                ActivityLevelEnrichmentStrategy(),
                TimestampEnrichmentStrategy()
            ]
        else:
            self.strategies = strategies
    
    def enrich(self, event: Dict[str, Any]) -> Dict[str, Any]:
        enriched = event
        for strategy in self.strategies:
            enriched = strategy.enrich(enriched)
        return enriched

